File: api/api/receituarios.js
// api/receituarios.js
// Função serverless Vercel — gera o Excel preenchido com dados do paciente
// usando ExcelJS, preservando formatação e fórmulas do modelo original.

const ExcelJS = require('exceljs');
const path = require('path');
const fs = require('fs');

// Caminho do modelo Excel dentro da pasta api/
const MODELO_PATH = path.join(__dirname, 'modelo.xlsm');

module.exports = async function handler(req, res) {
  // CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Método não permitido' });
  }

  try {
    const { nome, prontuario, maeNome, idade, sexo, data } = req.body || {};

    // Verifica se o modelo existe
    if (!fs.existsSync(MODELO_PATH)) {
      return res.status(500).json({
        error: 'Modelo Excel não encontrado',
        path: MODELO_PATH,
        dir: fs.readdirSync(__dirname),
      });
    }

    // Carrega o workbook preservando fórmulas e formatação
    const wb = new ExcelJS.Workbook();
    await wb.xlsx.readFile(MODELO_PATH);

    // Preenche apenas a aba Cadastro — as outras abas buscam via fórmula
    const ws = wb.getWorksheet('Cadastro');
    if (!ws) {
      return res.status(500).json({ error: 'Aba Cadastro não encontrada no modelo' });
    }

    ws.getCell('C7').value = (nome || '').toUpperCase();
    ws.getCell('C8').value = String(prontuario || '');
    ws.getCell('C9').value = (maeNome || '').toUpperCase();
    ws.getCell('C12').value = idade ? Number(idade) : '';
    ws.getCell('C13').value = sexo || '';
    ws.getCell('C14').value = 'GERIATRIA';
    ws.getCell('C16').value = data || '';

    // Gera o buffer do arquivo
    const buffer = await wb.xlsx.writeBuffer();

    const nomeArquivo = `Receituarios_${(nome || 'paciente').replace(/[^a-zA-Z\u00C0-\u00FF0-9 ]/g, '').trim()}.xlsx`;

    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
    res.setHeader('Content-Disposition', `attachment; filename="${nomeArquivo}"`);
    res.setHeader('Content-Length', buffer.length);
    return res.status(200).send(Buffer.from(buffer));

  } catch (err) {
    console.error('Erro ao gerar Excel:', err);
    return res.status(500).json({
      error: err.message,
      stack: err.stack,
    });
  }
};

File: api/receituarios.py
from http.server import BaseHTTPRequestHandler
import json, io, os, base64, traceback


def _carregar_modelo():
    base = os.path.dirname(__file__)
    p = os.path.join(base, "modelo.xlsm")
    if os.path.exists(p):
        with open(p, "rb") as f:
            return f.read()
    p2 = os.path.join(base, "modelo_b64.txt")
    if os.path.exists(p2):
        with open(p2, "r") as f:
            return base64.b64decode(f.read().strip())
    raise FileNotFoundError("modelo.xlsm nao encontrado em: " + base)


def _gerar(nome, prontuario, mae, idade, sexo, data):
    from openpyxl import load_workbook
    wb = load_workbook(io.BytesIO(_carregar_modelo()), keep_vba=True)
    ws = wb["Cadastro"]
    ws["C7"].value = (nome or "").upper()
    ws["C8"].value = str(prontuario or "")
    ws["C9"].value = (mae or "").upper()
    ws["C12"].value = int(idade) if str(idade).strip().isdigit() else (idade or "")
    ws["C13"].value = sexo or ""
    ws["C14"].value = "GERIATRIA"
    ws["C16"].value = data or ""
    buf = io.BytesIO()
    wb.save(buf)
    return buf.getvalue()


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_POST(self):
        try:
            n = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(n)) if n else {}
            arq = _gerar(
                body.get("nome", ""),
                body.get("prontuario", ""),
                body.get("maeNome", ""),
                body.get("idade", ""),
                body.get("sexo", ""),
                body.get("data", ""),
            )
            nome_arq = "Receituarios_" + (body.get("nome") or "paciente").replace(" ", "_") + ".xlsm"
            self.send_response(200)
            self._cors()
            self.send_header("Content-Type", "application/vnd.ms-excel.sheet.macroEnabled.12")
            self.send_header("Content-Disposition", 'attachment; filename="' + nome_arq + '"')
            self.send_header("Content-Length", str(len(arq)))
            self.end_headers()
            self.wfile.write(arq)
        except Exception as e:
            err = json.dumps({"error": str(e), "trace": traceback.format_exc()}).encode()
            self.send_response(500)
            self._cors()
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(err)))
            self.end_headers()
            self.wfile.write(err)

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def log_message(self, fmt, *args):
        pass
